lib/pbench/server/api/resources/query_apis/iteration_samples.py

```diff
@@ -0,0 +1,240 @@
+        means the authenticated user is the owner of the dataset, or has ADMIN
+        role.)
+
+        If the user has authorization to read the dataset, return the Dataset
+        object as CONTEXT so that the postprocess operation can proceed to get all
+        the iteration aggregations.
+        """
+        run_id = client_json.get("run_id")
+
+        # Query the dataset using the given run id
+        dataset = Dataset.query(md5=run_id)
+        if not dataset:
+            self.logger.error("Run ID {} not found", run_id)
+            abort(HTTPStatus.NOT_FOUND, message="Dataset not found")
+
+        owner = User.query(id=dataset.owner_id)
+        if not owner:
+            self.logger.error(
+                "Dataset owner ID {} cannot be found in Users", dataset.owner_id
+            )
+            abort(HTTPStatus.INTERNAL_SERVER_ERROR, message="Dataset owner not found")
+
+        # For Iteration samples, we check authorization against the ownership of the
+        # dataset that was selected rather than having an explicit "user"
+        # JSON parameter. This will raise UnauthorizedAccess on failure.
+        self._check_authorization(owner.username, dataset["access"])
+
+        # The dataset exists, and authenticated user has enough access so continue
+        # the operation with the appropriate CONTEXT.
+        return {"dataset": dataset}
+
+    def assemble(self, json_data: JSON, context: CONTEXT) -> JSON:
+        """
+        Construct a pbench search query for aggregating iteration samples of given run id
+        within a specified date range and which are either owned by a specified username,
+        or have been made publicly accessible.
+
+        {
+            "start": "start-time",
+            "end": "end-time",
+            "run_id": "run id string"
+        }
+
+        json_data: JSON dictionary of type-normalized parameters
+            "start" and "end" are datetime objects representing a set of Elasticsearch
+                run document indices in which to search.
+
+            "run_id": String representation of run id to match for aggregate all the iteration samples
+        """
+        start = json_data.get("start")
+        end = json_data.get("end")
+        run_id = json_data.get("run_id")
+
+        # We need to pass string dates as part of the Elasticsearch query; we
+        # use the unconverted strings passed by the caller rather than the
+        # adjusted and normalized datetime objects for this.
+        start_arg = f"{start:%Y-%m-%d}"
+        end_arg = f"{end:%Y-%m-%d}"
+
+        self.logger.info(
+            "Return iterations for user {}, prefix {}: ({} - {}) for run id: {}",
+            user,
+            self.prefix,
+            start,
+            end,
+            run_id,
+        )
+
+        uri_fragment = self._gen_month_range("result-data-sample", start, end)
+        self.logger.info("fragment, {}", uri_fragment)
+        return {
```

Comment:
As we discussed, we need to loop on the `scroll_id` until we have all of the `[hits][hits]` elements collected.

**Also**, the dashboard `service/dashboard.js` code takes a *list* of `run_id` values, and returns all of them as a list of the raw (scrolled) Elasticsearch results.

We can decide to implement a single `run_id` query (which is easier on us but will require the dashboard to make a separate server call for each `run_id` in the list), or we can replace the entire service by taking a list of `run_id` values and searching for each one.

(I'm actually tempted to say we should stick to the single `run_id` here, but I want to point this out so we think about it deliberately.)

Reply:
I'm not fully grokking this stuff yet, but I'm a little concerned about us collecting together the results of multiple scrolled calls and then returning it to the dashboard:  the ES results are returned piecemeal in order to avoid timeouts of the request to ES...if the Pbench Server aggregates those results, aren't we in danger of timing out the Dashboard's request to us?  And there's a similar concern regarding space:  the other reason for scrolled results is to avoid having the response grow too large; should we be concerned about a query which results in more entries than we can return in a single response?

I'm wondering if we aren't starting to solve a problem that we don't have (yet)...shouldn't we be waiting until the client has a discernible performance problem before we go implementing batching?

Reply:
Your concern is right Webb, we could end up with a query that's bigger than what we can return in single response. At the moment what the dashboard is doing is that it is going over all the iteration samples list again and aggregating the samples on top of all the aggregation returned from elasticsearch query. We could possibly do all that aggregation with elasticsearch query (by specifying more aggregation parameter) or perform the same aggregation on the server and only return the aggregated result. 
The problem is we don't fully understand the iteration samples postprocessing on the dashboard right now, so I am trying to set a meeting with Aquib and Anisha to discuss the what sort of aggregation we need to perform on the dashboard and if we can return only the aggregation that would be great.

Reply:
I neglected to actually *post* the earlier reply I made to this thread, but I'll add some more.

Yeah, it'd be nice if we could form a single *aggregation* that would include everything that the dashboard needs. If that's practical, we'd have a single Elasticsearch round-trip and latency wouldn't be an issue.

The original plan (represented in my architecture document) had been to limit the Elasticsearch dependency by moving both the dashboard service and model code into the server. However, as Nikhil says, the dashboard model code is complicated... and as Webb points out, moving all the Elasticsearch trips into the server poses a risk of exceeding client timeouts, and may not be a great idea. (We could consider building an asynchronous operation and an infrastructure to allow notifying the client when it's done, but that's a lot of infrastructure.)

The simplest immediate resolution would be to let the dashboard service layer continue to make repeated calls to the server, with each one fetching another scroll page of Elasticsearch data which the dashboard would consolidate as it does now. That's a bit messy, but a bit less messy than some of the other possible solutions.

Reply:
> The simplest immediate resolution would be to let the dashboard service layer continue to make repeated calls to the server, with each one fetching another scroll page of Elasticsearch data which the dashboard would consolidate as it does now. That's a bit messy, but a bit less messy than some of the other possible solutions.

So that means either sending back the elasticsearch scroll id or sending some sort of signal that indicate _there is another page of result possible would you like to scan_ and if the server receives `True` for the signal we scroll the next page (keeping the scroll_id in memory). We also have to make sure the scroll id doesnt expire in the meantime.

Reply:
The ES documentation offers [an alternative to using scrolling](https://www.elastic.co/guide/en/elasticsearch/reference/current/paginate-search-results.html#search-after) (which only works on queries of less than 10,000 documents anyway).  We might be able to build something on that, instead.

Reply:
That's intriguing, although I'm not entirely sure it helps. We probably don't need a `PIT` as we don't expect the result documents to change during the search, but we'd still need to retain the "search after" across the multiple client calls, so it's probably not really any easier than using the scroll ID. (Although without the PIT, I suppose we do dispense with that awkward lifetime issue, I'm not sure that the 1 minute interval imposes a massive hardship on the dashboard and it lets *us* throw away our internal storage -- PIT or scroll ID -- so we don't inherit the same lifetime problem that motivated Elasticsearch to add the `scroll=timelimit` syntax.)

Reply:
I grant that it's not much _easier_, but it's supposed to actually **work** (for large document-sets), which is plus.  😁 

If we can do without a `PIT` (which I, like you, suspect we can), then things get easier for everyone.  For starters, we're not racing against time, which is very nice.  And, if we can agree on the handshake, then we can let the client track the "search after" point and just hand it back to the server (e.g., as "opaque context"), so the server doesn't need to worry about keeping it and cleaning it up.